### ISSUE TITLE
Union[x, y] - false positive

### ISSUE THREAD

**Issue by SHxKM** [microsoft/pyright#1029]
**Describe the bug**
When `typing.Union` is used, upright raises a false-positive.

**To Reproduce**
```
def _get_date_if_string(datetime_string: Union[datetime.datetime, str]):
    if isinstance(datetime_string, datetime.datetime):
        return datetime_string
    return datetime.datetime.strptime(datetime_string, "%Y-%m-%dT%H:%M:%S.%fZ").replace(
        tzinfo=pytz.UTC
    )

def some_caller(asof: str):
    _get_date_if_string(datetime_string=asof)
```

**Expected behavior**
No warning.

**VS Code extension or command-line**
VS Code extension 1.1.70

**Additional context**
Thanks for a great library.

**Comment by erictraut** [comment]
What error or warning are you receiving? Pyright doesn't give me any error or warning with the code you provided. (Of course, I needed to add a few missing import statements.)

```
import datetime
import pytz
from typing import Union

def _get_date_if_string(datetime_string: Union[datetime.datetime, str]):
    if isinstance(datetime_string, datetime.datetime):
        return datetime_string
    return datetime.datetime.strptime(datetime_string, "%Y-%m-%dT%H:%M:%S.%fZ").replace(
        tzinfo=pytz.UTC
    )

def some_caller(asof: str):
    _get_date_if_string(datetime_string=asof)
```

**Comment by SHxKM** [comment]
Thanks for the quick response and sorry about missing those imports. Here's the error I'm getting:

![Screen Shot 2020-09-14 at 00 27 40](https://user-images.githubusercontent.com/6573474/93029084-43cbba00-f621-11ea-8c8d-c3471527c6a3.png)

If I changed the parameter `asof` in some caller to also say `Union[datetime.datetime, str]`, the error goes away. I guess it makes some sense? It's weird you're not seeing the error though.

**Comment by erictraut** [comment]
Are you sure that you have posted the exact code that is generating the error? I'm not able to reconcile the error message with the code. That particular error message should be used only in the case where a value is assigned to a variable that has a declared type. I don't see any variable assignment statements (i.e. `<name> = <expression>`) in your code sample, so I'm a bit mystified. Could you please try copying the exact code that I posted above and see if you can still repro it?

I also can't tell from your screen shot which expression is highlighted for this error. If you're able to still repro the problem with the code I posted above, please take a screen shot of that entire code block along with the output of the "Problems" panel.

**Comment by K3UL** [comment]
Hi,
I am seeing something that I think is related, please excuse me and tell me if this needs to be another issue. When using a Union, it seems like PyRight verifies that all the types in the union satisfy the code that follows.

It is actually bothering me with my JSON processings. Example : 

```python
# See https://github.com/python/typing/issues/182#issuecomment-186684288
JSONType = Union[None, bool, int, float, str, List[Any], Dict[str, Any]]

def create_series_from_json(payload: JSONType) -> pd.Series:
    index: List[dt.datetime] = []
    values: List[float] = []

    idx_times = payload["columns"].index("timestamp")
    idx_values = payload["columns"].index("value")
    for item in payload["items"]:
        index.append(_parse_time(item[idx_times]))
        values.append(item[idx_values])

    return pd.Series(values, index)
```

Here, unless I replace `JSONType` with `Dict[str, Any]`, I get : 
- For the lines with `payload["columns"].index()` : `cannot access member "index" for type bool`
- For the loop : `bool is not iterable`, `int` is not iterable, `float` is not iterable
And some other errors.

I think this might be by design (although extremely inconvenient) but I wanted to check because the way I understand Union, as long as one type fits, it should be OK.

**Comment by erictraut** [comment]
@K3UL, this is by design. Your understanding of Union is partly correct. Their behavior depends on whether they are the target or the source type when assigning a value. Let's look at a few examples.

```python
StrOrInt = Union[str, int]

def func(p1: StrOrInt):
   # In these examples, the union is the target of an assignment
   v1: StrOrInt = 3 # Fine
   v2: StrOrInt = "hi" # Fine
   v3: StrOrInt = p1 # Fine

   # In these examples, the union is the source of an assignment
   v4: str = p1 # Error
   v5: int = p1 # Error
```
In the example above, your function `create_series_from_json` claims that it can accept a bunch of different types (`None`, `bool`, `int`, etc.). But the implementation clearly doesn't handle most of those types. It will crash if someone passes `None` to the `payload` parameter. That's what the type checker is telling you in this case. To fix this, you need to either change the type of `payload` to be `Dict[str, str]` or fix the implementation of the function so it can handle all of the types within the `JSONType` union.

**Comment by K3UL** [comment]
Yes that is what I was afraid of. In my case it is very hard to define the type because it has a lot of nested levels.  Which does not make a lot of sense to me because in the end, different keys have values of different types, so the implementation can be correct depending on the key I am actually reading.  
I guess I am in for some more verbose modeling !

Thanks for your fast and comprehensive answer. I have been following the project for some time and I find inspiring how you carry it well !

**Comment by SHxKM** [comment]
@erictraut Here's the screenshot:

<img width="578" alt="Screen Shot 2020-09-14 at 22 23 12" src="https://user-images.githubusercontent.com/6573474/93129002-0aef1c00-f6d9-11ea-8ed7-31e55ce95cd4.png">

I *think* after I've read your comments to @K3UL, I understand this is marked by design. So basically even if the annotation is in the parameter, changing the type in the function is considered an error. For example this:

```
def create_subscription_from_alert(*, alert_id: str, asof: str) -> int:
    new_date = _get_date_if_string(datetime_string=asof) # asof not reassigned as datetime
```

Isn't marked as an error. 

> I have been following the project for some time and I find inspiring how you carry it well !

I have to second this.

**Comment by erictraut** [comment]
Thanks for the additional information. It's now clear to me what's happening.

The code in the screen shot has declared that the local symbol `asof` is a `str`. The function `_get_date_if_string` returns a value that is of type `datetime`, and you are attempting to assign that value to `asof`.

The correct (and easy!) workaround is to use a new symbol name rather than re-using the name `asof`.

```python
asof_date = _get_date_if_string(datetime_string=asof)
```

**Comment by bombsimon** [comment]
What's the preferred way to work with list of unions when you're fine with any variant? I can't get it to work even when explicitly defining the type. Example:

```python
str_or_int: list[str | int] = [1, "2", 3]

si0 = str_or_int[0]             # Error
si1: str | int = str_or_int[1]  # Error 
print(str_or_int[2])            # Error
```
